### DISCUSSION TITLE
AM Handler and permissible functions

### DISCUSSION THREAD

**Discussion by vchuravy** [openucx/ucx#6407] [Q&A] [answered]
Is there a set of functions that is permissible to be called from within an active message handler?
https://github.com/openucx/ucx/blob/2c81364f39cad13d06f2778ed8056d66cdba6e7f/src/ucp/api/ucp_def.h#L620-L623 explicitly disallows `ucp_worker_progress`, but are there any other?

What about concurrent accesses when the worker is initialized with `UCX_THREAD_MODE_MULTI`? Is it safe to call `ucp_worker_progress` from another thread?

I am asking since I just encountered a scenario where I called `ucp_worker_progress` from within the am handler because I was using concurrent tasks in Julia  and I switched from the handler to the progress task...

**Comment by yosefe** [comment, marked as answer]
- Only ucp_worker_progress and functions which may destroy the relevant objects (ep/worker/context) are not allowed
- When the worker is initialized with `UCX_THREAD_MODE_MULTI` it's safe to call `ucp_worker_progress ` from multiple threads

**Reply by vchuravy** [reply]
Thanks that helps a lot.
